src/jit/assertionprop.cpp

```diff
@@ -2629,9 +2608,29 @@ GenTreePtr Compiler::optConstantAssertionProp(AssertionDsc* curAssertion,
             else
             {
                 bool isArrIndex = ((tree->gtFlags & GTF_VAR_ARR_INDEX) != 0);
-                newTree->ChangeOperConst(GT_CNS_INT);
-                newTree->gtIntCon.gtIconVal = curAssertion->op2.u1.iconVal;
-                newTree->ClearIconHandleMask();
+                // If we have done constant propagation of a struct type, it is only valid for zero-init,
+                // and we have to ensure that we have the right zero for the type.
+                if (varTypeIsStruct(tree))
```

Comment:
Nit: this could simply be `assert(!varTypeIsStruct(tree) || (curAssertion->op2.u1.iconVal == 0))`